pkg/provider/provider_test.go

```diff
@@ -0,0 +1,839 @@
+					EntryPoints: []string{"http-10000"},
+					Priority:    1001,
+					Service:     "my-ns-svc-d-8080",
+				},
+				"my-ns-svc-e-8080": {
+					Rule:        "Host(`svc-e.my-ns.maesh`) || Host(`10.10.16.1`)",
+					EntryPoints: []string{"http-10000"},
+					Priority:    1001,
+					Service:     "my-ns-svc-e-8080",
+				},
+			},
+			Services: map[string]*dynamic.Service{
+				"block-all-service": blockAllService,
+				"readiness":         readinessSvc,
+				"my-ns-svc-b-8080": {
+					LoadBalancer: &dynamic.ServersLoadBalancer{
+						Servers: []dynamic.Server{
+							{URL: "https://10.10.2.1:8080"},
+							{URL: "https://10.10.2.2:8080"},
+						},
+						PassHostHeader: getBoolRef(true),
+					},
+				},
+				"my-ns-svc-b-ts-8080-traffic-split": {
+					Weighted: &dynamic.WeightedRoundRobin{
+						Services: []dynamic.WRRService{
+							{
+								Name:   "my-ns-svc-b-ts-8080-svc-d-traffic-split-backend",
+								Weight: getIntRef(80),
+							},
+							{
+								Name:   "my-ns-svc-b-ts-8080-svc-e-traffic-split-backend",
+								Weight: getIntRef(20),
+							},
+						},
+					},
+				},
+				"my-ns-svc-b-ts-8080-svc-d-traffic-split-backend": {
+					LoadBalancer: &dynamic.ServersLoadBalancer{
+						Servers: []dynamic.Server{
+							{URL: "https://svc-d.my-ns.maesh:8080"},
+						},
+						PassHostHeader: getBoolRef(false),
+					},
+				},
+				"my-ns-svc-b-ts-8080-svc-e-traffic-split-backend": {
+					LoadBalancer: &dynamic.ServersLoadBalancer{
+						Servers: []dynamic.Server{
+							{URL: "https://svc-e.my-ns.maesh:8080"},
+						},
+						PassHostHeader: getBoolRef(false),
+					},
+				},
+				"my-ns-svc-d-8080": {
+					LoadBalancer: &dynamic.ServersLoadBalancer{
+						PassHostHeader: getBoolRef(true),
+					},
+				},
+				"my-ns-svc-e-8080": {
+					LoadBalancer: &dynamic.ServersLoadBalancer{
+						PassHostHeader: getBoolRef(true),
+					},
+				},
+			},
+			Middlewares: map[string]*dynamic.Middleware{
+				"block-all-middleware": blockAllMiddleware,
+				"my-ns-svc-b": {
+					Retry: &dynamic.Retry{Attempts: 2},
+					RateLimit: &dynamic.RateLimit{
+						Average: 100,
+						Burst:   110,
+					},
+					CircuitBreaker: &dynamic.CircuitBreaker{
+						Expression: "LatencyAtQuantileMS(50.0) > 100",
+					},
+				},
+			},
+		},
+		TCP: &dynamic.TCPConfiguration{
+			Routers: map[string]*dynamic.TCPRouter{
+				"my-ns-svc-f-8080": {
+					EntryPoints: []string{"tcp-5000"},
+					Service:     "my-ns-svc-f-8080",
+					Rule:        "HostSNI(`*`)",
+				},
+			},
+			Services: map[string]*dynamic.TCPService{
+				"my-ns-svc-f-8080": {
+					LoadBalancer: &dynamic.TCPServersLoadBalancer{
+						Servers: []dynamic.TCPServer{
+							{
+								Address: "10.10.2.1:8080",
+							},
+							{
+								Address: "10.10.2.2:8080",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	assert.Equal(t, want, got)
+}
+
+func TestProvider_BuildConfigTCP(t *testing.T) {
+	saA := "sa-a"
```

Comment:
These tests are very difficult to review/maintain 😅 I'm not sure what to propose to make it better while keeping the same depth in testing though! 🤔

Reply:
Yes, I'm worried about that too. In order to test the provider we need to generate a graph, which take space and is difficult to follow. Then we have to compare the resulting dynamic config to the expected one. 

We could eventually marshal the expected dynamic config and put it in a JSON file. It could remove a lot of code, but I'm not sure it would make it more maintainable.
Then, for the topology, we could have a serializer and deserializer, and put them in a JSON file.

WDYT?

Reply:
Sounds good to me, might be easier to understand indeed :)

Reply:
Instead of providing a serializer for the Topology, I'm seriously considering removing the circular references stuff in favour of an "index by Key" model, like in Traefik's dynamic config. This would make the structure a little less convenient to use but much more resilient and extendable.

Reply:
Done in 7cc784e
It makes tests much more maintainable. I also took the opportunity to add more tests.